In [1]:
#Paquetes importados

import networkx as nx
from networkx.algorithms.community import label_propagation_communities
import networkx.algorithms.community as nx_comm
import openpyxl
import pandas as pd

df = pd.DataFrame()

## Análisis previo

### Lectura de la red de estudio

In [2]:
G0 = nx.read_pajek("1919.net")
G0 = G0.to_undirected()

### Descomposición en comunidades con un algoritmo de Lovaina

In [3]:
c0 = nx_comm.louvain_communities(G0)

### Evalúo la modularidad de esa partición en comunidades

In [4]:
q0=nx_comm.modularity(G0, c0)
q0

0.3478952896973566

#### El valor de la modularidad, para este caso, es prácticamente igual al obtenido mediante Gephi.

### Obtengo la lista de nodos de la componente conexa más grande

In [5]:
giant0 = max(nx.connected_components(G0), key=len)

### Obtengo el subgrafo consistente SOLO en la componente gigante

In [6]:
S0 = G0.subgraph(max(nx.connected_components(G0), key=len)).copy()

In [7]:
# A veces hay aristas repetidas y networkx lee el archivo como
# "multigrafo", para el cual no funciona k-cores. Así que genero
# acá una copia simplificada.
Gsimp0 = nx.Graph()
for u,v,data in G0.edges(data=True):
    if not Gsimp0.has_edge(u,v):
        Gsimp0.add_edge(u, v)

In [8]:
c00 = nx_comm.louvain_communities(Gsimp0)
q00=nx_comm.modularity(Gsimp0, c00)
q00

0.35677108744544755

### Veo el tamaño del k-core y su profundidad

In [9]:
k0=len(nx.k_core(Gsimp0))
kcore1=len(nx.k_core(Gsimp0,1))
kcore2=len(nx.k_core(Gsimp0,2))
kcore28=len(nx.k_core(Gsimp0,28))
kcorek0=len(nx.k_core(Gsimp0,k0))
k0

31

In [10]:
Num_nodos_0=[kcore1,kcore2,kcore28,kcorek0]
Num_nodos_0

[291, 285, 31, 0]

####  A pesar de no ser la misma cantidad de nodos por capa, en la primera capa podemos visualizar 291 nodos y obtenemos la misma profundidad de k-cores, puesto que en el 31-core no tengo nodos por ende la profundidad es 30. 

### Calculo el coeficiente de correlación de Pearson

####  La asotividad mide la similitud de las conexiones en el gráfico con respecto al grado del nodo, tendremos dos opciones para calcular este coeficiente: degree_assortativity_coefficient - degree_pearson_correlation_coefficient. Puesto que son practicamente el mismo valor, solo difieren en la función potencial usaremos por simplicidad la deasortatividad directo.

In [11]:
nx.degree_pearson_correlation_coefficient(nx.k_core(Gsimp0,1))

0.17747559928596435

In [12]:
nx.degree_pearson_correlation_coefficient(G0)

0.17747559928596435

In [13]:
nx.degree_pearson_correlation_coefficient(Gsimp0)

0.17747559928596435

In [14]:
nx.degree_assortativity_coefficient(G0)

0.17747559928596335

In [15]:
Coef_corr_0=nx.degree_assortativity_coefficient(Gsimp0)
Coef_corr_0

0.17747559928596335

### Calculo el coeficiente de clustering

In [16]:
C_0=nx.average_clustering(Gsimp0)

#### El valor del coeficiente de clustering, para este caso, es prácticamente igual al obtenido mediante Gephi.


## Generación de 10 redes aleatorizadas

### Se generan 10 redes aleatorizadas con la misma distribución de grados

In [17]:
G_1 = nx.configuration_model([d for v, d in G0.degree()])
G_2 = nx.configuration_model([d for v, d in G0.degree()])
G_3 = nx.configuration_model([d for v, d in G0.degree()])
G_4 = nx.configuration_model([d for v, d in G0.degree()])
G_5 = nx.configuration_model([d for v, d in G0.degree()])
G_6 = nx.configuration_model([d for v, d in G0.degree()])
G_7 = nx.configuration_model([d for v, d in G0.degree()])
G_8 = nx.configuration_model([d for v, d in G0.degree()])
G_9 = nx.configuration_model([d for v, d in G0.degree()])
G_10 = nx.configuration_model([d for v, d in G0.degree()])

### Calculamos la modularidad de cada una de ellas tanto para red original como la de componente gigante

In [18]:
c1 = nx_comm.louvain_communities(G_1)
c2 = nx_comm.louvain_communities(G_2)
c3 = nx_comm.louvain_communities(G_3)
c4 = nx_comm.louvain_communities(G_4)
c5 = nx_comm.louvain_communities(G_5)
c6 = nx_comm.louvain_communities(G_6)
c7 = nx_comm.louvain_communities(G_7)
c8 = nx_comm.louvain_communities(G_8)
c9 = nx_comm.louvain_communities(G_9)
c10 = nx_comm.louvain_communities(G_10)

q1 = nx_comm.modularity(G_1,c1)
q2 = nx_comm.modularity(G_2,c2)
q3 = nx_comm.modularity(G_3,c3)
q4 = nx_comm.modularity(G_4,c4)
q5 = nx_comm.modularity(G_5,c5)
q6 = nx_comm.modularity(G_6,c6)
q7 = nx_comm.modularity(G_7,c7)
q8 = nx_comm.modularity(G_8,c8)
q9 = nx_comm.modularity(G_9,c9)
q10 = nx_comm.modularity(G_10,c10)


#### Lista de nodos de la componente conexa más grande

In [19]:
giant_1 = max(nx.connected_components(G_1), key=len)
Gsimp_1 = nx.Graph()
for u,v,data in G_1.edges(data=True):
    if not Gsimp_1.has_edge(u,v):
        Gsimp_1.add_edge(u, v)

In [20]:
giant_2 = max(nx.connected_components(G_2), key=len)
Gsimp_2 = nx.Graph()
for u,v,data in G_2.edges(data=True):
    if not Gsimp_2.has_edge(u,v):
        Gsimp_2.add_edge(u, v)


In [21]:
giant_3 = max(nx.connected_components(G_3), key=len)
Gsimp_3 = nx.Graph()
for u,v,data in G_3.edges(data=True):
    if not Gsimp_3.has_edge(u,v):
        Gsimp_3.add_edge(u, v)


In [22]:
giant_4 = max(nx.connected_components(G_4), key=len)
Gsimp_4 = nx.Graph()
for u,v,data in G_4.edges(data=True):
    if not Gsimp_4.has_edge(u,v):
        Gsimp_4.add_edge(u, v)


In [23]:
giant_5 = max(nx.connected_components(G_5), key=len)
Gsimp_5 = nx.Graph()
for u,v,data in G_5.edges(data=True):
    if not Gsimp_5.has_edge(u,v):
        Gsimp_5.add_edge(u, v)


In [24]:
giant_6 = max(nx.connected_components(G_6), key=len)
Gsimp_6 = nx.Graph()
for u,v,data in G_6.edges(data=True):
    if not Gsimp_6.has_edge(u,v):
        Gsimp_6.add_edge(u, v)


In [25]:
giant_7 = max(nx.connected_components(G_7), key=len)
Gsimp_7 = nx.Graph()
for u,v,data in G_7.edges(data=True):
    if not Gsimp_7.has_edge(u,v):
        Gsimp_7.add_edge(u, v)


In [26]:
giant_8 = max(nx.connected_components(G_8), key=len)
Gsimp_8 = nx.Graph()
for u,v,data in G_8.edges(data=True):
    if not Gsimp_8.has_edge(u,v):
        Gsimp_8.add_edge(u, v)


In [27]:
giant_9 = max(nx.connected_components(G_9), key=len)
Gsimp_9 = nx.Graph()
for u,v,data in G_9.edges(data=True):
    if not Gsimp_9.has_edge(u,v):
        Gsimp_9.add_edge(u, v)


In [28]:
giant_10 = max(nx.connected_components(G_10), key=len)
Gsimp_10 = nx.Graph()
for u,v,data in G_10.edges(data=True):
    if not Gsimp_10.has_edge(u,v):
        Gsimp_10.add_edge(u, v)


In [29]:
c_1 = nx_comm.louvain_communities(Gsimp_1)
c_2 = nx_comm.louvain_communities(Gsimp_2)
c_3 = nx_comm.louvain_communities(Gsimp_3)
c_4 = nx_comm.louvain_communities(Gsimp_4)
c_5 = nx_comm.louvain_communities(Gsimp_5)
c_6 = nx_comm.louvain_communities(Gsimp_6)
c_7 = nx_comm.louvain_communities(Gsimp_7)
c_8 = nx_comm.louvain_communities(Gsimp_8)
c_9 = nx_comm.louvain_communities(Gsimp_9)
c_10 = nx_comm.louvain_communities(Gsimp_10)

q_1 = nx_comm.modularity(Gsimp_1,c_1)
q_2 = nx_comm.modularity(Gsimp_2,c_2)
q_3 = nx_comm.modularity(Gsimp_3,c_3)
q_4 = nx_comm.modularity(Gsimp_4,c_4)
q_5 = nx_comm.modularity(Gsimp_5,c_5)
q_6 = nx_comm.modularity(Gsimp_6,c_6)
q_7 = nx_comm.modularity(Gsimp_7,c_7)
q_8 = nx_comm.modularity(Gsimp_8,c_8)
q_9 = nx_comm.modularity(Gsimp_9,c_9)
q_10 = nx_comm.modularity(Gsimp_10,c_10)


### Calculamos el coeficiente de correlación de Newman

#### Red completa

In [30]:
Coef_new_1=nx.degree_assortativity_coefficient(G_1)
Coef_new_2=nx.degree_assortativity_coefficient(G_2)
Coef_new_3=nx.degree_assortativity_coefficient(G_3)
Coef_new_4=nx.degree_assortativity_coefficient(G_4)
Coef_new_5=nx.degree_assortativity_coefficient(G_5)
Coef_new_6=nx.degree_assortativity_coefficient(G_6)
Coef_new_7=nx.degree_assortativity_coefficient(G_7)
Coef_new_8=nx.degree_assortativity_coefficient(G_8)
Coef_new_9=nx.degree_assortativity_coefficient(G_9)
Coef_new_10=nx.degree_assortativity_coefficient(G_10)

#### Red componente gigante

In [31]:
Coef_corr_1=nx.degree_assortativity_coefficient(Gsimp_1)
Coef_corr_2=nx.degree_assortativity_coefficient(Gsimp_2)
Coef_corr_3=nx.degree_assortativity_coefficient(Gsimp_3)
Coef_corr_4=nx.degree_assortativity_coefficient(Gsimp_4)
Coef_corr_5=nx.degree_assortativity_coefficient(Gsimp_5)
Coef_corr_6=nx.degree_assortativity_coefficient(Gsimp_6)
Coef_corr_7=nx.degree_assortativity_coefficient(Gsimp_7)
Coef_corr_8=nx.degree_assortativity_coefficient(Gsimp_8)
Coef_corr_9=nx.degree_assortativity_coefficient(Gsimp_9)
Coef_corr_10=nx.degree_assortativity_coefficient(Gsimp_10)

### Calculamos el coeficiente de clustering local

In [32]:
C_1=nx.average_clustering(Gsimp_1)
C_2=nx.average_clustering(Gsimp_2)
C_3=nx.average_clustering(Gsimp_3) 
C_4=nx.average_clustering(Gsimp_4)
C_5=nx.average_clustering(Gsimp_5)
C_6=nx.average_clustering(Gsimp_6)
C_7=nx.average_clustering(Gsimp_7)
C_8=nx.average_clustering(Gsimp_8) 
C_9=nx.average_clustering(Gsimp_9)
C_10=nx.average_clustering(Gsimp_10)

In [33]:
#Paso para no generar error en la obtencion k-cores
Gsimp_1.remove_edges_from(nx.selfloop_edges(Gsimp_1))
Gsimp_2.remove_edges_from(nx.selfloop_edges(Gsimp_2))
Gsimp_3.remove_edges_from(nx.selfloop_edges(Gsimp_3))
Gsimp_4.remove_edges_from(nx.selfloop_edges(Gsimp_4))
Gsimp_5.remove_edges_from(nx.selfloop_edges(Gsimp_5))
Gsimp_6.remove_edges_from(nx.selfloop_edges(Gsimp_6))
Gsimp_7.remove_edges_from(nx.selfloop_edges(Gsimp_7))
Gsimp_8.remove_edges_from(nx.selfloop_edges(Gsimp_8))
Gsimp_9.remove_edges_from(nx.selfloop_edges(Gsimp_9))
Gsimp_10.remove_edges_from(nx.selfloop_edges(Gsimp_10))

### Calculamos la profundidad de k-cores

In [34]:
kcore_1=len(nx.k_core(Gsimp_1))
kcore_2=len(nx.k_core(Gsimp_2))
kcore_3=len(nx.k_core(Gsimp_3))
kcore_4=len(nx.k_core(Gsimp_4))
kcore_5=len(nx.k_core(Gsimp_5))
kcore_6=len(nx.k_core(Gsimp_6))
kcore_7=len(nx.k_core(Gsimp_7))
kcore_8=len(nx.k_core(Gsimp_8))
kcore_9=len(nx.k_core(Gsimp_9))
kcore_10=len(nx.k_core(Gsimp_10))



## Generar Tabla de contenido:

In [35]:
Modularidad_Original=[q0,q1,q2,q3,q4,q5,q6,q7,q8,q9,q10]
Modularidad_Gigante=[q00,q_1,q_2,q_3,q_4,q_5,q_6,q_7,q_8,q_9,q_10]
Coef_Newman=[Coef_corr_0,Coef_corr_1,Coef_corr_2,Coef_corr_3,Coef_corr_4,Coef_corr_5,Coef_corr_6,Coef_corr_7,Coef_corr_8,Coef_corr_9,Coef_corr_10]
Coef_Clustering=[C_0,C_1,C_2,C_3,C_4,C_5,C_6,C_7,C_8,C_9,C_10]
Profundidad_k_core=[k0,kcore_1,kcore_2,kcore_3,kcore_4,kcore_5,kcore_6,kcore_7,kcore_8,kcore_9,kcore_10]
df['Modularidad_Original'] = Modularidad_Original
df['Modularidad_Gigante'] = Modularidad_Gigante
df['Coef_Newman'] = Coef_Newman
df['Coef_Clustering'] = Coef_Clustering
df['Profundidad_k_core'] = Profundidad_k_core

print(df)

    Modularidad_Original  Modularidad_Gigante  Coef_Newman  Coef_Clustering  \
0               0.347895             0.356771     0.177476         0.682333   
1               0.175288             0.174461    -0.032097         0.193257   
2               0.172907             0.169600    -0.041811         0.190781   
3               0.174376             0.168735    -0.047700         0.190120   
4               0.169908             0.172852    -0.070753         0.188190   
5               0.175657             0.169064    -0.045852         0.178994   
6               0.178212             0.171336    -0.039512         0.186234   
7               0.167080             0.165021    -0.067328         0.192113   
8               0.176309             0.170355    -0.071062         0.190527   
9               0.180199             0.168927    -0.030639         0.190058   
10              0.171108             0.168575    -0.054839         0.189556   

    Profundidad_k_core  
0                   31  
1

In [36]:
file_name = 'Redes_aleatorizadas.xlsx'
df.to_excel(file_name)
